components/opc-ua/src/Microsoft.Azure.IIoT.OpcUa.Edge.Publisher/src/Engine/LegacyJobOrchestrator.cs

```diff
@@ -522,87 +581,54 @@ public class LegacyJobOrchestrator : IJobOrchestrator, IPublisherConfigServices,
                 }
 
-                foreach (var entry in existingGroup) {
-                    if (!entry.OpcNodes.Any()) {
-                        _publishedNodesEntrys.Remove(entry);
-                    }
-                }
+                updatedPublishedNodesEntries = updatedPublishedNodesEntries.Where(entry => entry.OpcNodes.Count > 0).ToList();
 
-                var found = false;
-                await _lockJobs.WaitAsync(ct).ConfigureAwait(false);
-                try {
-                    var jobs = _publishedNodesJobConverter.ToWriterGroupJobs(existingGroup, _legacyCliModel);
-                    if (jobs.Any()) {
-                        foreach (var job in jobs) {
-                            var newJob = ToJobProcessingInstructionModel(job);
-                            if (string.IsNullOrEmpty(newJob?.Job?.Id)) {
-                                continue;
-                            }
-                            foreach (var assignedJob in _assignedJobs) {
-                                if (newJob.Job.Id == assignedJob.Value.Job.Id) {
-                                    _assignedJobs[assignedJob.Key] = newJob;
-                                    found = true;
-                                    break;
-                                }
-                            }
-                            if (!found) {
-                                if (_availableJobs.ContainsKey(newJob.Job.Id)) {
-                                    _availableJobs[newJob.Job.Id] = newJob;
-                                    found = true;
-                                }
-                            }
-                        }
-                    }
+                _publishedNodesEntries.Clear();
+                _publishedNodesEntries.AddRange(updatedPublishedNodesEntries);
 
-                    if (!found) {
-                        var entryJobId = _publishedNodesJobConverter.
-                            ToConnectionModel(request, _legacyCliModel).CreateConnectionId();
-                        foreach (var assignedJob in _assignedJobs) {
-                            if (entryJobId == assignedJob.Value.Job.Id) {
-                                found = _assignedJobs.Remove(assignedJob.Key, out _);
-                                break;
-                            }
-                        }
-                        if (!found) {
-                            found = _availableJobs.Remove(entryJobId, out _);
-                        }
-                    }
-                }
-                finally {
-                    _lockJobs.Release();
-                }
+                PersistPublishedNodes();
 
-                if (!found) {
-                    throw new MethodCallStatusException((int)HttpStatusCode.NotFound, "Endpoint not found.");
-                }
+                // Update local state.
+                RefreshJobs(updatedPublishedNodesEntries);
```

Comment:
>RefreshJobs [](http://example.com/codeflow?start=16&length=11)

Overall, I'm not convinced this approach is better, faster or more readable  than the existing implementation. Furthermore the publish/unpublish are now no longer symmetric.
I like the optimizations with the opc nodes hashset and simplification of dataset check.

Reply:
OK. I am reverting back to original job refresh mechanism that was present here.

Reply:
Done.